src/main/java/com/aws/greengrass/util/platforms/windows/WindowsExec.java

```diff
@@ -118,6 +124,76 @@ public synchronized void close() throws IOException {
             return;
         }
 
+        try {
+            stopGracefully();
+        } finally {
+            stopForcefully();
+        }
+    }
+
+    @SuppressFBWarnings("SWL_SLEEP_WITH_LOCK_HELD")
+    private void stopGracefully() {
+        //TODO: Temporarily limiting graceful shutdown to only processes that are created using ProcessImplForWin32
+        if (!(process instanceof ProcessImplForWin32)) {
+            return;
+        }
+        int pid = process instanceof ProcessImplForWin32
+                ? ((ProcessImplForWin32) process).getPid() : Processes.newPidProcess(process).getPid();
+        boolean sentConsoleCtrlEvent = false;
+        synchronized (Kernel32.INSTANCE) {
+            Kernel32 k32 = Kernel32.INSTANCE;
+            // if the console is already attached, a second attach call will fail.
+            if (!k32.FreeConsole()) {
+                logger.error("FreeConsole error {}", k32.GetLastError());
+            }
+            if (!k32.AttachConsole(pid)) {
+                logger.error("AttachConsole error {}", k32.GetLastError());
+                // Console already attached so we cannot signal it
+                if (k32.GetLastError() == ERROR_ACCESS_DENIED) {
+                    logger.info("AttachConsole failed for PID: {}, calling FreeConsole on it", pid);
+                    // Oddly, calling FreeConsole seems to make subsequent calls work?
+                    if (!k32.FreeConsole()) {
+                        logger.error("FreeConsole error {}", k32.GetLastError());
+                    }
+                }
+                return;
+            }
+            Kernel32Ex k32Ex = Kernel32Ex.INSTANCE;
+            try {
+                if (!k32Ex.SetConsoleCtrlHandler(null, true)) {
+                    logger.error("SetConsoleCtrlHandler add error {}", k32.GetLastError());
+                    return;
+                }
+                if (k32.GenerateConsoleCtrlEvent(Wincon.CTRL_C_EVENT, 0)) {
+                    sentConsoleCtrlEvent = true;
+                } else {
+                    logger.error("GenerateConsoleCtrlEvent error {}", k32.GetLastError());
+                }
+            } finally {
+                if (!k32.FreeConsole()) {
+                    logger.error("FreeConsole error {}", k32.GetLastError());
+                }
+                // wait to ensure CtrlHandler is not enabled before the calling process gets the signal
+                try {
+                    Thread.sleep(2000);
```

Comment:
Does it need to be 2 seconds? Is that the only time you tried?

Reply:
Do we need these timeouts (the 5s and 2s here) to be configurable?
If 2 seconds is not long enough, what happens?
If it causes a crash, we'll get customer tickets

Reply:
Not likely. The 5 second timeout is the same as on linux which should be configurable eventually, but I wouldn't do it right now.

I don't see how Windows would need more than 2 seconds to send a signal, but who knows.

Reply:
ok - lets make sure we have a backlog to address this (configurating the timeout) then. 

I know @fahadmohammed01  tried 1 second and was still seeing the error occur so he switched to 2

Reply:
wow...